### PR TITLE
Recparser update, new attempt

### PR DESCRIPTION
**Pull request by Logg-y**
Recreation of #56 with just my commits in, without the 50+ commits that happened over the past three weeks.

Continuation of #56 

----

Quite a few hours of staring with a hex editor later:

1) I worked out roughly how the recorded game file is structured, and use that internal tree to navigate instead of searching semi-blindly for things. This should hopefully be more reliable in the long run, and in the meantime makes figuring out more about the file format a lot easier.
2) I worked out how to parse the commands list - this does get exposed on the RecordedGameMetadata object but I am deleting it before saving to Mongo because it gets VERY large. I did try to implement a tidy-ish TS interface to all the actions, including ones that aren't really very useful at the moment.
3) I implemented a XMB parser, which can read the embedded game data files. The Aoe3de resource manager (that's no longer on github) was a huge help and saved me from having to work out the format myself.

This adds:
 * RecordedGameMetadata
   * `commands` - provides both "crude" and "refined" lists of game commands (this object gets very large and is deleted before saving to DB)
   * `commandParserError` - if parsing commands fails for any reason, the error is written here. This *does not constitute an overall parser failure*, but all values derived from the commands list will just not be populated.
   * `commandParserWarnings` - certain unusual things found when parsing the commands list (commands with parts that couldn't be interpreted correctly or that look like they are illegal) write warnings here. Stripped out before being saved to DB.
   * `commandParserGeneratedWarnings` - Boolean value for whether `commandParserWarnings` contained anything. This IS saved to DB, in the future it might be possible to query for it to find recs and investigate any anomalies it finds
   * `unresignedPlayers` - a list of player IDs that did not resign. This could be used to infer the winner of the game, though it will fail if a player is outright defeated by killing all their stuff.
 * RecordedGamePlayerMetadata
   *  `civName` - the major god name. This is now resolved from the packed game data at parse time and will be static. The IDs _probably_ won't change either, but using this in preference to the ids will mean that is no longer a risk. Use of this would mean that unknown/new/modded major gods can be displayed with their internal name instead of "unknown <number>" too.
   * `techTimes` - A mapping of tech names to the last time the player attempted to do something to research them. This is unreliable as there's no way to tell from the recorded game if the research attempt actually finished - it might never actually start due to not having enough resources, get cancelled by the player, or the building doing the research getting destroyed. Currently not being saved to DB as its value is questionable - but getting minor gods means doing nearly all the work for this anyway.
   * `classicalAgeMinorGod`, `heroicAgeMinorGod`, `mythicAgeMinorGod` - Minor god techs, as determined by the packed game data, with names adjusted to remove the age prefix from the tech names (eg "Hermes" "Bragi" "Hekate"). Like `techTimes` this simply contains the last minor god tech of each age the player tried to research, there is no way to tell from the recorded game if it went through.
   * `approximateCommandsPerMinute` - This one seemed like a little bit of fun and was pretty easy to derive once figuring out the command list and the internal action IDs - it's not completely accurate, and it certainly isn't comparable to raw APM as it ignores everything like selecting units.
   * `rLinkId` - The player's official API user profile ID.

I did not expose it, but included in this PR is a function that will parse the map's initial object list, mostly included because I had worked out where it was and how to parse it. This could in future potentially permit things like:
 * Identify map variants, eg whether ghost lake starts with 1 or 2 gold mines based on resource proximity to each player's TC
 * Probably way too much work to be realistic (and has processing/storage implications), but this is a step towards generating an image of the actual game's minimap at the start of the game rather than using the loading screen images from the game assets. I suspect that this might be technically possible to do, though.

For now I added the new properties to the Mongo schema except `commands` (due to its size), `techTimes`, and the raw warnings list `commandParserWarnings`. These all get stripped out before saving to the DB in the interests of not packing tons of info that's not useful.

### CODE REVIEW

**Comment by jdduprey** [comment]
minor gods!!!

File: src/components/recs/team-tile.tsx [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+import { majorGodIndexToData, majorGodNameToData } from "@/types/MajorGods";
 import { IRecordedGame } from "@/types/RecordedGame";
-import { RecordedGamePlayerMetadata } from "@/types/RecordedGameParser";
+import { RecordedGamePlayerMetadata } from "@/types/recParser/RecordedGameParser";
 import Image from "next/image";
 
 function PlayerTile(playerData: RecordedGamePlayerMetadata, gameGuid: string) {
   const { name, civ, id } = playerData;
-  const godData = majorGodIndexToData(civ);
+  let godData = majorGodIndexToData(civ);
```
Comment (erin-fitzpatric) [inline src/components/recs/team-tile.tsx]: Why was this changed to a `let`? Is this getting changed somewhere that I am not seeing?

File: src/components/recs/team-tile.tsx [hunk 2]
```diff
@@ -1,12 +1,13 @@
 import { teamIndexToPlayerData } from "@/server/teams";
-import { majorGodIndexToData } from "@/types/MajorGods";
+import { majorGodIndexToData, majorGodNameToData } from "@/types/MajorGods";
```
Comment (erin-fitzpatric) [inline src/components/recs/team-tile.tsx]: Is majorGodNameToData being used?

File: src/db/mongo/schemas/recordedGame/RecordedGameSchema.ts [hunk 3]
```diff
@@ -26,6 +26,8 @@ const RecordedGameSchema = new Schema(
     version: { type: Number, required: false, default: 0},
     gameLength: {type: Number, required: false, default: 0},
     teams: { type: [[Number]]},
+    commandParserError: { type: String, required: false, default: "" },
+    unresignedPlayers: { type: [[Number]], required: false },
```
Comment (erin-fitzpatric) [inline src/db/mongo/schemas/recordedGame/RecordedGameSchema.ts:31]: What would the value of unresignedPlayers be? Are you thinking this will be used to determine the winner of this match without doing a match history lookup from the Athens API?

File: src/server/controllers/upload-rec-controller.ts [hunk 4]
```diff
@@ -16,6 +16,9 @@ export function mapRecGameMetadata(data: RecordedGameMetadata) {
   mappedData.playerData = mappedData.playerData.filter(
     (_player, idx) => idx !== 0
   );
+  delete mappedData.commands;
+  delete mappedData.commandParserWarnings;
+  mappedData.playerData.forEach((playerData) => delete playerData.techTimes);
```
Comment (erin-fitzpatric) [inline src/server/controllers/upload-rec-controller.ts:29]: Can you elaborate on what this is doing?

File: src/components/recs/team-tile.tsx (hunk 1 above)
Comment (Logg-y) [reply src/components/recs/team-tile.tsx]: I missed this when reverting the frontend change here - I'll sort this

File: src/components/recs/team-tile.tsx (hunk 2 above)
Comment (Logg-y) [reply src/components/recs/team-tile.tsx]: I also forgot to undo this when undoing the frontend display change - will sort this as well

File: src/db/mongo/schemas/recordedGame/RecordedGameSchema.ts (hunk 3 above)
Comment (Logg-y) [reply src/db/mongo/schemas/recordedGame/RecordedGameSchema.ts:31]: It's an array of player numbers who (at least according to the rec data) didn't resign. It could absolutely be used to work out winners without quering the API, but will fail if a player was actually eliminated (by killing all their stuff) rather than resigning before that.

File: src/server/controllers/upload-rec-controller.ts (hunk 4 above)
Comment (Logg-y) [reply src/server/controllers/upload-rec-controller.ts:29]: At the moment these are fields that are part of the parser output but probably aren't useful for the recs page, though if anyone (admittedly if I'm being realistic it's probably only ever going to be me!) ends up using the parser code for other purposes then returning this stuff can be of interest. So far I've only really had reason to want to do that twice, to look at exactly what orders caused unexpected stuff to happen.

So this just gets rid of the fields - if Mongoose does this on its own to things outside the schema it's probably unnecessary, but I was not sure if that was the case or not.

Maybe it's better to just never add these to the returned objects or something? I'm not too sure.

**Comment by Logg-y** [comment]
Hopefully addressed the above!

File: src/components/recs/team-tile.tsx (hunk 1 above)
Comment (erin-fitzpatric) [reply src/components/recs/team-tile.tsx]: let me know when you are able to update this

File: src/db/mongo/schemas/recordedGame/RecordedGameSchema.ts (hunk 3 above)
Comment (erin-fitzpatric) [reply src/db/mongo/schemas/recordedGame/RecordedGameSchema.ts:31]: I am not certain it is useful data if it only works in edge cases, but don't see any harm in storing the data. Thanks for the explanation.

File: src/server/controllers/upload-rec-controller.ts (hunk 4 above)
Comment (erin-fitzpatric) [reply src/server/controllers/upload-rec-controller.ts:29]: To your last point, the more elegant approach would be to use a Mongo pipeline aggregation and `project` the fields that are used by the front end services. I'm ok leaving what you have since I don't want to blow up the rec flow right now...but could be a future PR if you wanted to clean this up a bit.

**Review by erin-fitzpatric — COMMENTED** [review]
A few final comments - I swear we will get this merged this weekend!!!!

**Comment by Logg-y** [comment]
Wow, I'm so sorry it's taken me a 3rd swing to undo all that D:

_Hopefully_ sorted now. For reals?

**Review by erin-fitzpatric — APPROVED** [review]

**Comment by erin-fitzpatric** [comment]
@Logg-y - clean up the merge conflict and we can merge :)

**Comment by Logg-y** [comment]
Hopefully sorted!
